Applications/Spire/Source/Ui/RealSpinBox.cpp

```diff
@@ -0,0 +1,441 @@
+    m_has_first_click = true;
+  }
+  auto control = get_current_control(event->pos());
+  if(control == QStyle::SubControl::SC_ScrollBarAddLine) {
+    m_up_arrow_timer_id = startTimer(qApp->keyboardInputInterval());
+  } else if(control == QStyle::SubControl::SC_ScrollBarSubLine) {
+    m_down_arrow_timer_id = startTimer(qApp->keyboardInputInterval());
+  }
+  QAbstractSpinBox::mousePressEvent(event);
+}
+
+void RealSpinBox::mouseReleaseEvent(QMouseEvent* event) {
+  stop_timer(m_up_arrow_timer_id);
+  stop_timer(m_down_arrow_timer_id);
+  QAbstractSpinBox::mouseReleaseEvent(event);
+}
+
+void RealSpinBox::timerEvent(QTimerEvent* event) {
+  if(event->timerId() == m_up_arrow_timer_id) {
+    stop_timer(m_up_arrow_timer_id);
+    m_up_arrow_timer_id = startTimer(MOUSE_REPEAT_DELAY_MS);
+    add_step(1, qApp->keyboardModifiers());
+  } else if(event->timerId() == m_down_arrow_timer_id) {
+    stop_timer(m_down_arrow_timer_id);
+    m_down_arrow_timer_id = startTimer(MOUSE_REPEAT_DELAY_MS);
+    add_step(-1, qApp->keyboardModifiers());
+  }
+}
+
+QAbstractSpinBox::StepEnabled RealSpinBox::stepEnabled() const {
+  return QAbstractSpinBox::StepUpEnabled | QAbstractSpinBox::StepDownEnabled;
+}
+
+void RealSpinBox::set_decimal_places(int decimals) {
+  m_decimals = std::min(decimals, MAXIMUM_DECIMAL_PLACES);
+  auto point_count = [&] {
+    if(m_decimals > 0) {
+      return 1;
+    }
+    return 0;
+  }();
+  m_real_regex = QRegularExpression(QString(
+    "^[%1]?([0-9]*[%2]{0,%3}[0-9]{0,%4})$").arg(m_locale.negativeSign())
+    .arg(m_locale.decimalPoint()).arg(point_count).arg(m_decimals));
+}
+
+void RealSpinBox::set_minimum_decimal_places(int decimals) {
+  m_minimum_decimals = decimals;
+}
+
+void RealSpinBox::set_minimum(Real minimum) {
+  m_minimum = minimum;
+  auto value = get_value(text());
+  if(value != boost::none && value->compare(m_minimum) < 0) {
+    m_last_valid_value = m_minimum;
+    lineEdit()->setText(display_string(m_minimum));
+  }
+}
+
+void RealSpinBox::set_maximum(Real maximum) {
+  m_maximum = maximum;
+  auto value = get_value(text());
+  if(value != boost::none && value->compare(m_maximum) > 0) {
+    m_last_valid_value = m_maximum;
+    lineEdit()->setText(display_string(m_maximum));
+  }
+}
+
+void RealSpinBox::set_step(Real step) {
+  m_step = step;
+}
+
+
+RealSpinBox::Real RealSpinBox::get_value() const {
+  return m_last_valid_value;
+}
+
+void RealSpinBox::set_value(Real value) {
+  blockSignals(true);
+  value = clamped_real(value, m_minimum, m_maximum);
+  lineEdit()->setText(display_string(value));
+  blockSignals(false);
+}
+
+void RealSpinBox::stepBy(int step) {
+  add_step(step);
+}
+
+connection RealSpinBox::connect_change_signal(
+    const ChangeSignal::slot_type& slot) const {
+  return m_change_signal.connect(slot);
+}
+
+void RealSpinBox::add_step(int step) {
+  add_step(step, Qt::NoModifier);
+}
+
+void RealSpinBox::add_step(int step, Qt::KeyboardModifiers modifiers) {
+  step = [&] {
```

Comment:
Use IIFE only for initialization, in this case it's perfectly fine to do:

    if(condition) {
      step *= SHIFT_STEPS;
    }